d2layouts/d2sequence/layout.go

```diff
@@ -4,90 +4,140 @@ import (
 	"context"
 	"fmt"
 	"math"
+	"sort"
 
 	"oss.terrastruct.com/d2/d2graph"
 	"oss.terrastruct.com/d2/lib/geo"
 	"oss.terrastruct.com/d2/lib/go2"
 	"oss.terrastruct.com/d2/lib/label"
+	"oss.terrastruct.com/d2/lib/shape"
 )
 
 func Layout(ctx context.Context, g *d2graph.Graph) (err error) {
-	edgeYStep := MIN_EDGE_DISTANCE
-	actorXStep := MIN_ACTOR_DISTANCE
-	maxActorHeight := 0.
+	sd := &sequenceDiagram{
+		graph:          g,
+		objectRank:     make(map[*d2graph.Object]int),
+		objectDepth:    make(map[*d2graph.Object]int),
+		minEdgeRank:    make(map[*d2graph.Object]int),
+		maxEdgeRank:    make(map[*d2graph.Object]int),
+		edgeYStep:      MIN_EDGE_DISTANCE,
+		actorXStep:     MIN_ACTOR_DISTANCE,
+		maxActorHeight: 0.,
+	}
+
+	sd.init()
+	sd.placeActors()
+	sd.placeSpanBoxes()
+	sd.routeEdges()
+	sd.addLifelineEdges()
+
+	return nil
+}
+
+type sequenceDiagram struct {
+	graph *d2graph.Graph
+
+	edges  []*d2graph.Edge
+	actors []*d2graph.Object
+	spans  []*d2graph.Object
+
+	// can be either actors or span boxes
+	// rank: left to right position of actors/span boxes
+	objectRank map[*d2graph.Object]int
+	// depth: the nested levels of a given actor/span
+	objectDepth map[*d2graph.Object]int
```

Comment:
can you just use this? https://github.com/terrastruct/d2/blob/27a6e219a4c92726dfc40c05a3837076ed208cea/d2graph/d2graph.go#L386

Reply:
Indeed.. 😓

Reply:
Though, I'd rather have this here as `Level()` is a recursive call and here I just keep the state in ~3 places I need to use it

Reply:
nah code cleanliness over performance here, it's all negligible